src/app/clusters/commodity-price-server/commodity-price-server.cpp

```diff
@@ -141,11 +183,332 @@ void Instance::HandleGetDetailedForecastRequest(HandlerContext & ctx,
+    }
+
+    mCurrency = newValue;
+    if (oldValue != newValue)
+    {
+        ChipLogDetail(AppServer, "Endpoint: %d - mCurrency updated to Currency: %d DecimalPoints: %d", mEndpointId,
+                      mCurrency.currency, mCurrency.decimalPoints);
+        MatterReportingAttributeChangeCallback(mEndpointId, CommodityPrice::Id, Currency::Id);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::SetCurrentPrice(const DataModel::Nullable<Structs::CommodityPriceStruct::Type> newValue)
+{
+    // Check PeriodStart  < PeriodEnd (if not null)
+    if (!newValue.IsNull())
+    {
+        if (!newValue.Value().periodEnd.IsNull() && (newValue.Value().periodStart > newValue.Value().periodEnd.Value()))
+        {
+            return CHIP_IM_GLOBAL_STATUS(ConstraintError);
+        }
+
+        if (!newValue.Value().price.HasValue() && !newValue.Value().priceLevel.HasValue())
+        {
+            // Must have Price or PriceLevel
+            return CHIP_IM_GLOBAL_STATUS(ConstraintError);
+        }
+    }
+
+    // Do a deep copy of the newValue into mCurrentPrice
+    ReturnErrorOnFailure(CopyPrice(newValue));
+
+    ChipLogDetail(AppServer, "Endpoint: %d - mCurrentPrice updated", mEndpointId);
+    MatterReportingAttributeChangeCallback(mEndpointId, CommodityPrice::Id, CurrentPrice::Id);
+
+    // generate a PriceChange Event
+    GeneratePriceChangeEvent();
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::CopyStringToSpan(const char * src, Platform::ScopedMemoryBuffer<char> & bufferOut, CharSpan & spanOut)
+{
+    size_t len = strlen(src);
+    if (bufferOut.Get() != nullptr)
+    {
+        bufferOut.Free();
+    }
+
+    if (!bufferOut.Calloc(len))
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+    memcpy(bufferOut.Get(), src, len);
+    spanOut = CharSpan(bufferOut.Get(), len);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR Instance::CopyPrice(const DataModel::Nullable<Structs::CommodityPriceStruct::Type> & src)
+{
+    if (mCurrentPrice.IsNull())
+    {
+        // Check that the buffer for components should be nullptr
+        VerifyOrDie(mOwnedCurrentPriceComponentBuffer.Get() == nullptr);
+    }
+    else
+    {
+        // Free the .components
+        VerifyOrDie(mOwnedCurrentPriceComponentBuffer.Get() != nullptr);
+        mOwnedCurrentPriceComponentBuffer.Free();
+
+        // The .description is held within a ScopedBuffer so the
+        // CopyStringToSpan() will take care of that for us
+    }
+
+    // At this point we should have free'd all previous memory
+    if (src.IsNull())
+    {
+        mCurrentPrice.SetNull();
+    }
+    else
+    {
+        // Do a basic copy of the CommodityPriceStruct trivial fields
+        mCurrentPrice = src;
+
+        // Deep copy description (if present)
+        if (src.Value().description.HasValue())
+        {
+            CharSpan span;
+            CopyStringToSpan(src.Value().description.Value().data(), mOwnedCurrentPriceDescriptionBuffer, span);
```

Comment:
`src.Value().description.Value().data()` is not null-terminated, so this is a security bug.

You should just pass `src.Value().description.Value()` to your copying function, so you know the right length to copy.